[GSoC] (DO NOT MERGE) Adding Discrete-Time TransferFunction

leomanga opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
This PR lays the groundwork for extending the lti module to discrete time. It implements the `TransferFunctionBase` class, which is inherited by both `TransferFunction` and `DiscreteTransferFunction`. It also extends the ability to create interconnections between discrete-time systems.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.control
  * Added a new class for discrete time transfer functions.
  * Added a general method for creating both discrete-time and continuous-time transfer functions
<!-- END RELEASE NOTES -->

leomanga committed:
first changes to include discrete-time systems

add pretty printing for discrete-time transfer functions

remove unnecessary newline in at the end of the file

leomanga committed:
physics.control: Add tests

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -475,6 +484,356 @@ def test_TransferFunction_is_biproper():
 
+def test_DTTransferFunction_construction():
+    z = symbols("z")
+    tf = DTTransferFunction(z + 1, z**2 + z + 1, z, 10)
```
What do you plan for having fully symbolic transfer functions?

I've seen some recent activity about a pade approximation elsewhere in SymPy, are you going to use that?

leomanga replied:
What do you mean?
The Padé approximation can be useful for approximating time delays. Can it be used for other purposes as well?

moorepants replied:
I haven't taken digital control in a long long time, so I don't remember everything, but I thought that when transforming a continous transfer function into a discrete transfer function that you have a choice in the method of discretization. I thought using a pade approximation was one of them, but now I think I mis remembered.

So maybe my question should be: how are you going to manage different choices in discretization? Or is this assuming a zero order hold always?

leomanga committed:
test: Add tests for TransferFunctionBase, DTTransferFunction, StateSpaceBase in test_args

TODO: write test for DTStateSpace

leomanga committed:
physics.control: Add gbt, bilinear, forward_diff and backward diff in DTTransferFunction class

leomanga committed:
physics.control: Add tests for DTTransferFunction operations

leomanga committed:
Remove unnecessary blank lines

leomanga committed:
physics.control: add compatibility checks, add tests

leomanga committed:
physics.control: Update control documentation, add .rewrite(DTStateSpace)

leomanga committed:
physics.control: Add compatibility tests,  fix transfer function initialization in Series divisions and returned to previous eval_frequency calculation

leomanga committed:
physics.control: Extend TransferFunctionMatrix and MIMO interconnection to discrete-time.
Add tests.

leomanga committed:
Revert formatting for readability

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -475,6 +486,498 @@ def test_TransferFunction_is_biproper():
+    expect1_ = DTTransferFunction(2*s + 1, 5*s**2 + 4*s + 3, s)
+    assert SP1.subs({a0: 1, a1: 2, b0: 3, b1: 4, b2: 5}) == expect1_
+    assert SP1.subs({a0: 1, a1: 2, b0: 3, b1: 4, b2: 5}).evalf() == expect1
+    assert expect1_.evalf() == expect1
```
It doesn't seem that you check whether SP1 has the expected symbolic form, you only check by substituting numbers.

leomanga replied:
These tests are intended for `subs()` and `evalf()`. The checks in question are part of the construction section.

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -475,6 +486,498 @@ def test_TransferFunction_is_biproper():
+                 5*a2*b1*s**(S(8)/3)/3 + S(25)*b1*s/9) - \
+                    5*3**(S(1)/3)*5**(S(2)/3)*\
+                        b0/(9*a1*a2*s**(S(8)/3) + 15*a1*s + \
+                            15*a2*b1*s**(S(8)/3) + 25*b1*s)
```
Best to avoid `\` for line continuation. Just use parentheses around long lines.

leomanga replied:
> Do these expected results come from text book answers? How do you know they are correct results?

These tests are the same as the ones from the continuous-time case, so I assume they hold. I just added the one at line 722, which comes from here https://it.mathworks.com/help/control/ref/dynamicsystem.evalfr.html

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -475,6 +486,498 @@ def test_TransferFunction_is_biproper():
+    # Low-frequency (or DC) gain.
+    assert DTTransferFunction(z, z - 1, z).dc_gain() == oo
+    assert DTTransferFunction(z - 1, z**2 + 3*z + 1, z).dc_gain() == 0
+    assert DTTransferFunction(z, z + 2, z, 0.2).dc_gain() == Rational(1, 3)
```
Would be nice to know if these are textbook results.

leomanga replied:
I've applied the definition of dc_gain manually. I can find some references if needed.

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -1188,6 +2521,16 @@ def test_MIMOFeedback_errors():
 
+    dtfm_1 = TransferFunctionMatrix([[dtf1, dtf2], [dtf3, dtf4]])
+    dtfm_2 = TransferFunctionMatrix([[dtf2, dtf3], [dtf4, dtf1]])
+    dtfm_3 = TransferFunctionMatrix.from_Matrix(eye(2), var=s)
```
If this is a new method I would make `Matrix` all lowercase.

leomanga commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -17,19 +17,28 @@
+    raises(NotImplementedError,
+        lambda: TransferFunctionBase.from_coeff_lists([1], [1], s))
+    raises(NotImplementedError,
+        lambda: TransferFunctionBase.from_zpk([1], [1], 1, s))
```
I've done this because these functions must be inherited by a subclass, and we need to be aware that they exist.

leomanga commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -238,7 +249,7 @@ def test_TransferFunction_functions():
 
     # evaluate the transfer function at particular frequencies.
-    assert tf1.eval_frequency(wn) == wn**2/(wn**2 + 4*wn - 5) + 2*wn/(wn**2 + 4*wn - 5) - 3/(wn**2 + 4*wn - 5)
+    assert tf1.eval_frequency(wn) == wn/(wn + 5) + 3/(wn + 5)
```
I changed the `eval_frequency` calculation to avoid building an unnecessary class. The result is the same, but in this case it's just in simplified form.

moorepants commented on doc/src/modules/physics/control/control.rst:
```diff
@@ -4,17 +4,19 @@ Control
-property is used for transfer functions in the ``Parallel`` class, and the
-multiplicative property is used for transfer functions in the ``Series`` class.
+represent Continuous-time Transfer functions in the Laplace domain, while
+``DTTransferFunction`` represent Discrete-time Transfer functions in the
```
I think we should name this "DiscreteTranferFunction", i.e. use full words instead of single letters. I'm open for other options, but it is customary in Python to be descriptive with names.

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -438,69 +452,124 @@ def wrapper(*args, **kwargs):
+class TransferFunctionBase(SISOLinearTimeInvariant):
+    r"""
+    Base class for transfer tunction objects.
+    This class is not meant to be used directly.
```
if we don't want users to use it we can name it `_TransferFunction...`

leomanga replied:
Maybe users could use that class in `isinstance` to check if a system is a transfer function.
In addition that class should appear in the documentation because a lot of function are defined there.

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -615,16 +687,22 @@ class TransferFunction(SISOLinearTimeInvariant):
+            raise NotImplementedError(
+                """
+                The TransferFunctionBase class is not meant tobe used directly.
+                """)
```
Python has an AbstractBaseClass that manages such things. Consider using it.

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -633,27 +711,38 @@ def __new__(cls, num, den, var):
 
     @classmethod
-    def from_rational_expression(cls, expr, var=None):
+    def from_rational_expression(cls, expr, var=None, *args, **kwargs):
```
We generally prefer to avoid `*args, **kwargs` and to name all arguments explicitly (especially if it is a public class).

leomanga replied:
Because this function could be used by different transfer function classes, that’s the only solution I can think of.
Anyway, the discrete-time transfer function overrides that function to make the arguments explicit.
At this point, I could do the same with `TransferFunction`.

leomanga commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -475,6 +486,498 @@ def test_TransferFunction_is_biproper():
+    tf6_ = DTTransferFunction(1/2, 4, s)
+    assert tf6_.num == 0.5
+    assert tf6_.den == 4
+    assert tf6_.args == (0.500000000000000, 4, s, 1)
```
Are you referring to this?
```Python
import math
Float(0.1, 9).is_same(Float(0.1, 10), math.isclose)
```

leomanga committed:
physics.control: Rename DTTransferFunction and use ABC for base class

leomanga committed:
physics.control: Update test_args

leomanga commented on sympy/physics/control/lti.py:
```diff
@@ -759,6 +855,9 @@ def from_coeff_lists(cls, num_list, den_list, var):
         TransferFunction(p*s + 1, 2*p*s**2 + 4, s)
+        >>> dtf = DTTransferFunction.from_coeff_lists([2, 1, -3], [1, 4, 3, 2], z, sampling_time=0.1)
+        >>> dtf
+        DTTransferFunction(2*z**2 + z - 3, z**3 + 4*z**2 + 3*z + 2, z, 0.1)
```
In some cases, several methods are defined only in `TransferFunctionBase`. What’s the best way to avoid mixing discrete and continuous transfer functions in the examples?

leomanga committed:
physics.control: Improve docstrings for transfer function classes

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -1226,224 +1179,909 @@ def get_asymptotic_stability_conditions(self, cancel_poles_zeros = False):
+        See :func:`TransferFunctionBase.from_rational_expression`.
+        """
+        return super().from_rational_expression(expr, var,
+                                         sampling_time = sampling_time)
```
```suggestion
        return super().from_rational_expression(expr, var, sampling_time=sampling_time)
```

try not to have space around = on kwargs.

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -438,69 +453,124 @@ def wrapper(*args, **kwargs):
+                Incompatible systems detected.
+                All systems should be either continuous-time or discrete-time.
+                Found: {} and {}""".format(systems[0], system,))
+        if system.sampling_time != sampling_time:
```
Is this check robust for Float(), float(), Integer(), int(), Symbol, etc? I guess sampling time can be different types.

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -475,6 +486,498 @@ def test_TransferFunction_is_biproper():
+    raises(ValueError, lambda: (1 + p**4) - tf1)
+    raises(TypeError, lambda: tf5 + tf1)
+    raises(TypeError, lambda: tf5 - tf1)
+    raises(TypeError, lambda: (tf5 + tf6) + tf1)
+    raises(TypeError, lambda: (tf5 - tf6) - tf1)
```
It could be helpful to at least have a note explaining what you are testing e.g. "can't add transferfunctions with different sampling times".

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -1226,224 +1179,909 @@ def get_asymptotic_stability_conditions(self, cancel_poles_zeros = False):
-            raise ValueError("TransferFunction cannot be divided by {}.".
-                format(type(other)))
+    @classmethod
+    def gbt(cls, cont_tf, sampling_time, alpha, var):
```
This is not a good method name, I recommend making it something that is full words.

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -1226,224 +1179,909 @@ def get_asymptotic_stability_conditions(self, cancel_poles_zeros = False):
+    def gbt(cls, cont_tf, sampling_time, alpha, var):
+        r"""
+        Returns the discretized transfer function H(z) from a continuous
+        transfer function H(s).
```
For methods with args/kwargs besides self, make sure to have a parameters section.

leomanga committed:
physics.control: refactoring

leomanga committed:
physics.control: Add Parameters section in discretization methods

leomanga committed:
physics.control: Ensure sampling_time is sympified in DiscreteTransferFunction and DTStateSpace

leomanga committed:
physics.control: Rename _print_DTTransferFunction to _print_DiscreteTransferFunction

leomanga commented:
Regarding the release notes, is it better to wait for the next PRs where I will also implement the `DiscreteStateSpace`, or should I write a separate one for each feature?

moorepants commented:
> Regarding the release notes, is it better to wait for the next PRs where I will also implement the DiscreteStateSpace, or should I write a separate one for each feature?

You should add the release notes in this PR for the new features in this PR.

moorepants commented:
We discussed this in the meeting and Leonardo said he's worked out a textbook problem with the new classes, so I asked if he could add it as an example in the tutorial with the other controls examples to demonstrate that the new classes work as expected.

leomanga committed:
physics.control: Add Example in Tutorials and a first implementation of is_stable

leomanga committed:
Merge remote-tracking branch 'origin/master' into disc-time-systems

leomanga committed:
physics.control: Rename DTStateSpace to DiscreteStateSpace, refactoring

leomanga committed:
physics.control: Implement stability tests for DiscreteTransferFunction

leomanga committed:
physics.control: rename DTStateSpace to DiscreteStateSpace in test_args

leomanga committed:
physics.control: Update new_tf function and add error handling for DiscreteTransferFunction

leomanga committed:
physics.control: Remove unused imports in new_tf docstring example

leomanga commented:
@moorepants I made the `new_tf()` function public because I think it could be useful.
This PR should now be ready to merge.

I noticed some pre-existing style issues, but I don't think it's worth fixing them here since this PR is already quite large.
I’ve started working on the discrete counterpart of the state-space system and fixed some of those style problems there.

leomanga committed:
Merge remote-tracking branch 'origin/master' into disc-time-systems

leomanga committed:
physics.control.tests: Update expected values in DiscreteTransferFunction tests for a change in its form

moorepants commented on doc/src/tutorials/physics/control/control_problems.rst:
```diff
@@ -348,6 +348,79 @@ Solution
+Given
+
+        .. math::
+            G1(z) &= \frac{z}{z - \frac{1}{2}}\\
```
```suggestion
            G_1(z) &= \frac{z}{z - \frac{1}{2}}\\
```

moorepants commented on doc/src/tutorials/physics/control/control_problems.rst:
```diff
@@ -348,6 +348,79 @@ Solution
+        .. math::
+            G1(z) &= \frac{z}{z - \frac{1}{2}}\\
+
+            G2(z) &= \frac{z}{z - 1}\\
```
```suggestion
            G_2(z) &= \frac{z}{z - 1}\\
```

moorepants commented on doc/src/tutorials/physics/control/control_problems.rst:
```diff
@@ -348,6 +348,79 @@ Solution
+
+Solution
+
+        >>> # Imports
```
I don't think you need to indent the code.

moorepants commented on doc/src/tutorials/physics/control/control_problems.rst:
```diff
@@ -348,6 +348,79 @@ Solution
+Find
+
+1. The transfer function :math:`W(z)` from input :math:`r(k)` to output :math:`y(k)`.
+2. Determine a finite-duration input signal :math:`u(k)` (i.e., such that there exists an :math:`h \in \mathbb{N}` for which :math:`u(k) = 0` for all :math:`k \geq h`) so that the resulting forced response :math:`y_f(k)` converges to zero.
```
wrap to 79 characters

moorepants commented on sympy/physics/control/lti.py:
```diff
@@ -438,69 +488,163 @@ def wrapper(*args, **kwargs):
 
-    Generally, a dynamical system representing a physical model can be described in terms of Linear
-    Ordinary Differential Equations like -
+def new_tf(num, den, var, sampling_time):
```
`create_transfer_function` or something is more informative

leomanga commented on sympy/physics/control/lti.py:
```diff
@@ -1001,10 +1039,12 @@ def dc_gain(self):
         """
-        m = Mul(self.num, Pow(self.den, -1, evaluate=False), evaluate=False)
-        return limit(m, self.var, 0)
+        pass
```
It's because it's an abstract method

leomanga committed:
physics.control: Style changes

- Renamed `new_tf` to `new_transfer_function` for clarity
- Adjusted LaTeX and pretty printing outputs to use "sampling time" instead of "sampling_time".
- Adjusted tutorial formatting and changed the image type to svg

leomanga commented:
I made the proposed changes, when we are ready I will open a new PR clone, so we won't have problems with the png images I uploaded.

leomanga committed:
physics.control: Fix spacing in numbered list

leomanga committed:
physics.control: Fix formatting issues in control_problems.rst

oscarbenjamin commented on sympy/physics/control/lti.py:
```diff
@@ -1063,7 +1029,7 @@ def eval_frequency(self, other):
         >>> omega = 0.1
         >>> tf1.eval_frequency(I*omega)
-        1/(0.99 + 0.2*I)
+        0.970493088912852 - 0.196059209881384*I
```
SymPy expression do strange things with floats. Note that the original expression shown in the docstring does not evaluate to itself:
```python
In [1]: 1/(0.99 + 0.2*I)
Out[1]: 0.980296049406921⋅(0.99 - 0.2⋅ⅈ)
```

leomanga committed:
physics.control: Rename new_transfer_function to create_transfer_function and add tests

leomanga committed:
physics.control: Change image colors for better displaying and improve the Float comparison

leomanga committed:
physics.control: Refactor tests in test_DiscreteTransferFunction_functions
